🏗 Add APPROVERS.json to enable blocking bundle-size checks for extensions

danielrozenberg opened the pull request:
wip - not yet being used because the bundle-size app needs to be updated to use it

danielrozenberg commented on build-system/tasks/bundle-size/APPROVERS.json:
```diff
@@ -0,0 +1,18 @@
+{
```
This is just an example so we can get started. Later on we'll let teams opt-in to this

rcebulko replied:
Who will be the owners of this?

danielrozenberg replied:
For now let's leave the existing ownership model (wg-infra) - we can change it later

rcebulko replied:
I'm fine with wg-infra remaining the owner too

rcebulko commented on build-system/tasks/bundle-size/README.md:
```diff
@@ -0,0 +1,19 @@
+
+The `APPROVERS.json` file is a listing of compiled AMP runtime and extensions
+files, and for each file what teams are allowed to approve pull requests that
+increase to that file's bundle size (Brotli-compressed size) and what is the
```
`/increase to/increase/s`

rcebulko commented on build-system/tasks/bundle-size/README.md:
```diff
@@ -0,0 +1,19 @@
+The `APPROVERS.json` file is a listing of compiled AMP runtime and extensions
+files, and for each file what teams are allowed to approve pull requests that
+increase to that file's bundle size (Brotli-compressed size) and what is the
+threshold of an increase that requires special approval, in KBs.
```
This is a long sentence to parse. It might be clearer so say something like 

```markdown
The `APPROVERS.json` file is a listing of compiled AMP runtime and extensions files. Each rule has two parts:
- `approvers`:  blah blah blah
- `threshold`: bloo bloo bloo
```

rcebulko commented on build-system/tasks/bundle-size/README.md:
```diff
@@ -0,0 +1,19 @@
+and `{team2}`) will fall back to the set of default approvers, currently set to
+@ampproject/wg-runtime and @ampproject/wg-performance. Members of
+@ampproject/wg-infra can also approve all all bundle size increases or app
+failures, but this should only be done when the cause for infrastructure issues.
```
`when the cause for` typo?

danielrozenberg commented:
Addressed your comments. Please take another look, wanna get this right :)

As for who owns this: for now let's leave the existing ownership model (wg-infra) - we can change it later

rcebulko commented on build-system/tasks/bundle-size/README.md:
```diff
@@ -0,0 +1,24 @@
+  approve pull requests that increase the file's bundle size (Brotli-compressed
+  size)
+- `threshold`: number of KBs that the file can increase before requiring an
+  approval from the above team members.
```
Nit: I'd move the Brotli-compressed bit out of approvers and into threshold, ie.  `number of kilobytes by which the file's Brotli-compressed size may increase without require approval` or something of that nature

danielrozenberg replied:
Done

rsimha commented on build-system/tasks/bundle-size/APPROVERS.json:
```diff
@@ -0,0 +1,18 @@
+  },
+  "dist/v0/amp-story-1.0.js": {
+    "approvers": ["ampproject/wg-stories"],
+    "threshold": 0.5
```
Where is this file currently used? Do you think it's worth changing it to jsonc or json5 so it can support inline comments?

danielrozenberg replied:
I prefer using plain JSON so I won't have to pull in more dependencies in the bot. The README should provide enough context

rcebulko replied:
I would support json5 here

rsimha commented on build-system/tasks/bundle-size/README.md:
```diff
@@ -0,0 +1,23 @@
+The `APPROVERS.json` file is a listing of compiled AMP runtime and extensions
+files. Each rule has two parts:
+
+- `approvers`: list of GitHub project teams, members of which are allowed to
```
nit: Change...

> list of GitHub project teams, members of which are allowed to approve

... to...

> list of GitHub teams whose members may approve

danielrozenberg replied:
Done

rsimha commented on build-system/tasks/bundle-size/README.md:
```diff
@@ -0,0 +1,23 @@
+- `approvers`: list of GitHub project teams, members of which are allowed to
+  approve pull requests that fail the check
+- `threshold`: number of KBs that the file bundle size (Brotli-compressed size)
+  can increase before failing the check
```
nit: Change...

> number of KBs that the file bundle size (Brotli-compressed size) can increase

... to...

> number of kilobytes by which the brotli-compressed bundle size can increase"

danielrozenberg replied:
Done

rsimha commented on build-system/tasks/bundle-size/README.md:
```diff
@@ -0,0 +1,23 @@
+  can increase before failing the check
+
+Approval from any single member of any of the file's approval-teams is enough to
+satisfy the check.
```
nit: Is this sentence now implied by the reworded `approvers` text?

danielrozenberg replied:
Kinda, but there's nothing wrong with being extra explicit in documentation :)

rsimha commented on build-system/tasks/bundle-size/README.md:
```diff
@@ -0,0 +1,23 @@
+each require approval by a different set of teams (e.g., a PR that increases the
+bundle-size of three files, requiring approvals by `{team1}`, `{team1, team2}`,
+and `{team2}`) will fall back to the set of default approvers, currently set to
+@ampproject/wg-runtime and @ampproject/wg-performance. Members of
```
Does this happen when two different files increase in size? Or only when two files with different `approvers` lists increase in size?

Either way, I think the first sentence can be simplified. Something like:

> Note: If a PR requires a bundle-size approval (from more than one set of approvers) for more than one file, it will fall back to the default set of approvers (`@ampproject/wg-runtime and @ampproject/wg-performance`).

danielrozenberg replied:
Done

rsimha commented on build-system/tasks/bundle-size/README.md:
```diff
@@ -0,0 +1,23 @@
+bundle-size of three files, requiring approvals by `{team1}`, `{team1, team2}`,
+and `{team2}`) will fall back to the set of default approvers, currently set to
+@ampproject/wg-runtime and @ampproject/wg-performance. Members of
+@ampproject/wg-infra can also approve all all bundle size increases or app
```
nit: Remove extra "all"

> can also approve all all bundle size increases

danielrozenberg replied:
Done

rsimha commented on build-system/tasks/bundle-size/APPROVERS.json:
```diff
@@ -0,0 +1,18 @@
+{
+  "dist/v0.js": {
+    "approvers": ["ampproject/wg-performance", "ampproject/wg-runtime"],
+    "threshold": 0.1
```
For clarity, would it make sense to call this `thresholdKb`?

danielrozenberg replied:
I don't like this name though :(

rsimha replied:
Yeah, `thresholdKb` sounds goofy. I was thinking JS variable names like `delay_ms`. Documentation is good enough here.

rsimha commented on build-system/tasks/bundle-size/README.md:
```diff
@@ -0,0 +1,21 @@
+If a pull request requires a bundle-size approval from more than one _set_ of
+approvers, it will fall back to the default set of approvers
+(@ampproject/wg-runtime and @ampproject/wg-performance). Members of
+@ampproject/wg-infra can also approve all bundle size increases or app failures,
```
nit: Surround all working group names in backticks.

danielrozenberg replied:
Done
